### ISSUE TITLE
Should `can_connect` blindly return False when in the install hook?

### ISSUE THREAD

**Issue by sed-i** [canonical/operator#1033]
In a [recent itest](https://github.com/canonical/prometheus-k8s-operator/actions/runs/6421948537/job/17447234944?pr=537#step:7:526), I am seeing a surprising error (ops 2.7.0, juju 3.1.5).

Seems like the `can_connect` method returned True on install. Prometheus has a `can_connect` guard in the `_update_cert` method ([link](https://github.com/canonical/prometheus-k8s-operator/blob/bbb79a2e5f229372bcf8e8aa8350eb9497443829/src/charm.py#L402)), and the only way to reach the `pebble.exec` that failed in the traceback below, is past that guard. Sometimes the github runners are very slow so I imagine it could be possible the container being ready before juju had a chance to run on-install.

Just curious if we perhaps want `can_connect` to have some contextual awareness and somewhat artificially always return `can_connect -> False` when in the install/remove hook.

(The pebble error is [great](https://github.com/canonical/prometheus-rock/issues/33) BTW.)

```
unit-prometheus-k8s-0: 23:24:42.746 ERROR unit.prometheus-k8s/0.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/ops/pebble.py", line 1599, in _request_raw
    response = self.opener.open(request, timeout=self.timeout)
  File "/usr/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.8/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 400: Bad Request

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./src/charm.py", line 1000, in <module>
    main(PrometheusCharm)
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/ops/main.py", line 429, in main
    charm = charm_class(framework)
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/lib/charms/tempo_k8s/v0/charm_tracing.py", line 213, in wrap_init
    original_init(self, framework, *args, **kwargs)
  File "./src/charm.py", line 128, in __init__
    self._update_cert()
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/lib/charms/tempo_k8s/v0/charm_tracing.py", line 455, in wrapped_function
    return callable(*args, **kwargs)  # type: ignore
  File "./src/charm.py", line 444, in _update_cert
    self.container.exec(["update-ca-certificates", "--fresh"]).wait()
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/ops/model.py", line 2571, in exec
    return self._pebble.exec(
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/ops/pebble.py", line 2374, in exec
    resp = self._request('POST', '/v1/exec', body=body)
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/ops/pebble.py", line 1564, in _request
    response = self._request_raw(method, path, query, headers, data)
  File "/var/lib/juju/agents/unit-prometheus-k8s-0/charm/venv/ops/pebble.py", line 1610, in _request_raw
    raise APIError(body, code, status, message)
ops.pebble.APIError: cannot find executable "update-ca-certificates"
unit-prometheus-k8s-0: 23:24:43.034 ERROR juju.worker.uniter.operation hook "install" (via hook dispatching script: dispatch) failed: exit status 1
```

**Comment by sed-i** [comment]
BTW, should we add a `from e`? 

https://github.com/canonical/operator/blob/d7e15e82e38e2d7a27d20f48c0761fa60dc10542/ops/pebble.py#L1613

**Comment by tonyandrewmeyer** [comment]
> BTW, should we add a `from e`?
> 
> https://github.com/canonical/operator/blob/d7e15e82e38e2d7a27d20f48c0761fa60dc10542/ops/pebble.py#L1613

Imo, yes. I had noticed a few places that seem that they should have a `from` when raising. I was planning to go over that as part of #908.

**Comment by benhoyt** [comment]
Agreed that we should add `from e` here (and probably elsewhere) -- thanks @tonyandrewmeyer.

However, I'm a bit confused on the actual issue. It seems that `can_connect` returning `True` is the right thing here: Pebble is available.

Where is the `update-ca-certificates` executable coming from? Is the problem that the `install` hook is being fired earlier than we expect? (I think in general pebble-ready can happen at any time.) Is install being fired before K8s has properly set up the container? (That seems odd.)

So I want to understand better what's going on and what's expected here. But one possibility is adding to the guard in `_update_cert` a check that the executable exists too.

**Comment by sed-i** [comment]
> It seems that `can_connect` returning `True` is the right thing here: Pebble is available.

Yes, indeed, it seems that pebble is available, and ops is reflecting that correctly.
I am asking something different here:

I assume that pebble being ready during install (before pebble-ready event) would be surprising to most people.
Strictly speaking, charms shouldn't fail the install hook because of some expectation that pebble-ready to run only after install. But that brings us into [startup sequence](https://discourse.charmhub.io/t/new-juju-events-i-would-love-to-have/10952) considerations and essentially holistic charming a la common exit hook.
I am not opposed at all to common exit hook, but some are. And if you are, then blindly returning False from can_connect when in the install hook could bring in some sense into the startup sequence.

(`update-ca-certificates` was missing from that image, and is not related to the topic of this issue.)

**Comment by benhoyt** [comment]
Tony, Leon, and I discussed this on video, and here are a few notes I took.

- pebble-ready is guaranteed by Juju to be fired after install
- can_connect may or may not return True during install, potentially confusing charmers.
- all this is another ordering argument in favour of "common exit hook" or "holistic" or "common event code" style

Overall, however, Pebble can be ready at any time (and that's independent of the ordering Juju uses to fire pebble-ready) and `can_connect` shouldn't try to outsmart the reality of that, so we're closing this issue and will continue the discussion about event ordering and suchlike as we discuss holistic vs delta.
